### DISCUSSION TITLE
Custom error messages?

### DISCUSSION THREAD

**Discussion by Fampov** [redhat-developer/vscode-yaml#605] [Q&A]
I'm working on writing a JSON schema for my company's metadata files and I'm loving the extension and how it enforces the schema. One issue I'm encountering though is that the error messages for the properties are not very helpful (I'd prefer a message like "Name cannot be null"): 
![image](https://user-images.githubusercontent.com/23152866/135897192-45124430-ceae-4a27-9503-f3a17963ffc3.png)

That is a very simple, unimportant example but it becomes much more needed when the errors are the result of failures associated with regex patterns: 
![image](https://user-images.githubusercontent.com/23152866/135898432-b32380f7-4802-4a60-affa-37e2b0baa13e.png)

I am new to JSON schemas, so its very possible I've just overlooked how to properly configure them to be surfaced by the extension, but the few things I've tried that were suggested online don't seem to be captured and surfaced by the extension.

Does the extension support custom error messages within a JSON schema?

**Comment by gorkem** [comment]
I am not aware of a way to pass custom errors on JSON Schema specification. Is there a case beyond regex format validation that we would need a custom message?

**Reply by Fampov** [reply]
@gorkem regex is definitely the primary need because many of the people who will be working in the YAML won't be familiar with regex so errors like this won't mean anything to them. 

Beyond that, I think we can get away in other situations with the current behavior, but it would be nice to still be able to clarify things as I showed above. 

- "Name cannot be null" compared to "Incorrect type. Expected string" is just a little friendlier in my opinion and doesn't require any thinking. 
- Another example where it would be nice when an enum of options available. It would be nice to have a custom error that could share what types are available if its not a long list.
- Or the ability to specify why their value is invalid or the field is required beyond just that it is invalid would be nice.

**Comment by lucasrodes** [comment]
I second this. It would be great to customise the error message, especially for pattern-mismatch errors.

I found this: https://stackoverflow.com/a/45202473/5056599, unsure if we could use a similar structure.


I just realised that this can be achieved with `errorMessage` key! For instance:

```json
  "date_published": {
        "title": "Date of publication of the dataset",
        "description": "Date when the article was published.",
        "pattern": "(^\\d\\d\\d\\d\\-\\d\\d\\-\\d\\d$)|(^\\d{4}$)",
        "errorMessage": "Correct format of date_published is YYYY-MM-DD or YYYY.",
    }
```
